dev/tasks/build/builder.js

```diff
@@ -0,0 +1,217 @@
+ * For licensing, see LICENSE.md.
+ */
+
+/* jshint node: true */
+
+'use strict';
+
+var Builder;
+
+/**
+ * A CKEditor 5 release builder.
+ *
+ * @class Builder
+ */
+module.exports = Builder = function( target ) {
+	/**
+	 * The target directory where to create the build.
+	 *
+	 * **Warning**: if existing, this directory will be deleted before processing.
+	 *
+	 * @type {string}
+	 */
+	this.target = target || 'build';
+
+	/**
+	 * The temporary directory to use for build processing.
+	 *
+	 * **Warning**: if existing, this directory will be deleted before processing.
+	 *
+	 * @type {string}
+	 */
+	this.tmp = 'tmp';
+
+	/**
+	 * The list of tasks to be executed by the `build()` method. Each entry is an Array containing the name of the
+	 * method inside `tasks` to execute and the message to show to the end user when executing it.
+	 *
+	 * @type {Array}
+	 */
+	this.taskList = [
+		[ 'cleanup', 'Cleaning the "' + target + '" directory...' ],
+		[ 'copyToTmp', 'Copying source files for manipulation...' ],
+		[ 'removeAmdNamespace', 'AMD cleanup...' ],
+		[ 'optimize', 'Creating the optimized code...' ],
+		[ 'cleanupTmp', 'Removing the "' + this.tmp + '" directory...' ]
+	];
+};
+
+Builder.prototype = {
+	/**
+	 * Builds a CKEditor release based on the current development code.
+	 *
+	 * @param {Function} [callback] Function to be called when build finishes.
+	 */
+	build: function( callback ) {
+		var that = this;
+		var stepCounter = 0;
+
+		runNext();
+
+		function runNext() {
+			var next = that.taskList.shift();
+
+			if ( next ) {
+				stepCounter++;
+				console.log( stepCounter + '. ' + next[ 1 ] );
+				that.tasks[ next[ 0 ] ].call( that, runNext );
+			} else {
+				if ( callback ) {
+					callback();
+				}
+			}
+		}
+	},
+
+	/**
+	 * Holds individual methods for each task executed by the builder.
+	 *
+	 * All methods here MUST be called in the builder context by using
+	 * `builder.tasks.someMethod.call( builder, callback )`.
+	 */
+	tasks: {
+		/**
+		 * Deletes the `target` and `tmp` directories.
+		 *
+		 * @param {Function} callback Function to be called when the task is done.
+		 * @returns {Object} The callback returned valued.
+		 */
+		cleanup: function( callback ) {
+			var del = require( 'del' );
+			del.sync( this.target );
+			del.sync( this.tmp );
+
+			return callback();
+		},
+
+		/**
+		 * Copy the local source code of CKEditor and its dependencies to the `tmp` directory for processing.
+		 *
+		 * @param {Function} callback Function to be called when the task is done.
+		 * @returns {Object} The callback returned valued.
+		 */
+		copyToTmp: function( callback ) {
+			var ncp = require( 'ncp' ).ncp;
+			var path = require( 'path' );
+			var fs = require( 'fs' );
+			var tmp = this.tmp;
+
+			var deps = require( '../../../package.json' ).dependencies;
+
+			var toCopy = Object.keys( deps ).filter( function( name ) {
+				return name.indexOf( 'ckeditor5-' ) === 0;
+			} );
+
+			if ( !fs.existsSync( tmp ) ) {
```

Comment:
Since the `tmp/` directory is removed at the end of the task we risk removing someone's directory (with all his uncommitted work :D). Even if this is only a theoretical situation, it's a good practice IMO to abort if a directory which should not exist already exists.